kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/ProducerAppendInfo.java

```diff
@@ -0,0 +1,222 @@
+                                + "(incoming seq. number), %s (current end sequence number)",
+                        currentEntry.producerEpoch(), topicPartition, appendFirstSeq, currentLastSeq);
+                throw new OutOfOrderSequenceException(msg);
+            }
+        }
+    }
+
+    private Boolean inSequence(Integer lastSeq, Integer nextSeq) {
+        if (log.isDebugEnabled()) {
+            log.debug("check sequence lastSeq: {}, nextSeq: {}.", lastSeq, nextSeq);
+        }
+        return nextSeq == lastSeq + 1L || (nextSeq == 0 && lastSeq == Integer.MAX_VALUE);
+    }
+
+    public Optional<CompletedTxn> append(RecordBatch batch, Optional<Long> firstOffset) {
+        if (log.isDebugEnabled()) {
+            log.debug("Append batch: pid: {} baseSequence: {} lastSequence: {} baseOffset: {}  lastOffset: {} ",
+                    batch.producerId(), batch.baseSequence(), batch.lastSequence(), batch.baseOffset(),
+                    batch.lastOffset());
+        }
+        if (batch.isControlBatch()) {
+            Iterator<Record> recordIterator = batch.iterator();
+            if (recordIterator.hasNext()) {
+                Record record = recordIterator.next();
+                EndTransactionMarker endTxnMarker = EndTransactionMarker.deserialize(record);
+                return appendEndTxnMarker(
+                        endTxnMarker, batch.producerEpoch(), batch.baseOffset(), record.timestamp());
+            } else {
+                // An empty control batch means the entire transaction has been cleaned from the log,
+                // so no need to append
+                return Optional.empty();
+            }
+        } else {
+            appendDataBatch(batch.producerEpoch(), batch.baseSequence(), batch.lastSequence(), batch.maxTimestamp(),
+                    firstOffset.orElse(batch.baseOffset()), batch.lastOffset());
+            return Optional.empty();
+        }
+    }
+
+    public void appendDataBatch(Short epoch, Integer firstSeq, Integer lastSeq, Long lastTimestamp,
+                                Long firstOffset, Long lastOffset) {
+        if (log.isDebugEnabled()) {
+            log.debug("append data batch epoch: {}, firstSeq: {}, lastSeq: {}, firstOffset: {}, lastOffset: {}",
+                    epoch, firstSeq, lastSeq, firstOffset, lastOffset);
+        }
+        maybeValidateDataBatch(epoch, firstSeq);
+        // We use lastSeq - firstSeq instead of lastOffset - firstOffset, because current offset is wrong,
+        // the offset is set when message published to storage.
+        updatedEntry.addBatch(epoch, lastSeq, lastOffset, lastSeq - firstSeq, lastTimestamp);
+    }
+
+    /**
+     * When message publish success, we should save the current txn first offset.
+     *
+     * @param isTransactional is transaction or not
+     * @param firstOffset transaction first offset
+     */
+    public void updateCurrentTxnFirstOffset(Boolean isTransactional, long firstOffset) {
```

Comment:
I didn't find this method in Kafka 2.8.1. It looks like it's inside the `ProducerAppendInfo#appendDataBatch` method in Kafka.

Reply:
The reason why I add this method is that we can't get a real offset when doing `ProducerAppendInfo#appendDataBatch`. In KoP we must update the txn first offset after append message into the storage. Otherwise, the aborted index might be wrong.